enhancements/machine-api/control-plane-set.md

```diff
@@ -0,0 +1,585 @@
+}
+
+type ControlPlaneSetStrategy struct {
+    // Type defines the type of update strategy that should be
+    // used when updating Machines owned by the ControlPlaneSet.
+    // Valid values are "RollingUpdate", "Recreate" and "OnDelete".
+    // The current default value is "RollingUpdate".
+    // +kubebuilder:default:="RollingUpdate"
+    // +kubebuilder:validation:Enum:="RollingUpdate";"Recreate";"OnDelete"
+    // +optional
+    Type ControlPlaneSetStrategyType `json:"type,omitempty"`
+
+    // This is left as a struct to allow future rolling update
+    // strategy configuration to be added later.
+}
+
+// ControlPlaneSetStrategyType is an enumeration of different update strategies
+// for the Control Plane Machines.
+type ControlPlaneSetStrategyType string
+
+const (
+    // RollingUpdate is the default update strategy type for a
+    // ControlPlaneSet. This will cause the ControlPlaneSet to
+    // first create a new Machine and wait for this to be Ready
+    // before removing the Machine chosen for replacement.
+    RollingUpdate ControlPlaneSetStrategyType = "RollingUpdate"
+
+    // Recreate causes the ControlPlaneSet controller to first
+    // remove a ControlPlaneMachine before creating its
+    // replacement. This allows for scenarios with limited capacity
+    // such as baremetal environments where additional capacity to
+    // perform rolling updates is not available.
+    Recreate ControlPlaneSetStrategyType = "Recreate"
+
+    // OnDelete causes the ControlPlaneSet to only replace a
+    // Machine once it has been marked for deletion. This strategy
+    // makes the rollout of updated specifications into a manual
+    // process. This allows users to test new configuration on
+    // a single Machine without forcing the rollout of all of their
+    // Control Plane Machines.
+    OnDelete ControlPlaneSetStrategyType = "OnDelete"
+)
+
+// ControlPlaneSetStatus represents the status of the ControlPlaneSet CRD.
+type ControlPlaneSetStatus struct {
+    // ObservedGeneration is the most recent generation observed for this
+    // ControlPlaneSet. It corresponds to the ControlPlaneSets's generation,
+    // which is updated on mutation by the API Server.
+    // +optional
+    ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+    // Replicas is the number of Control Plane Machines created by the
+    // ControlPlaneSet controller.
+    // Note that during update operations this value may differ from the
+    // desired replica count.
+    Replicas int32 `json:"replicas"`
+
+    // ReadyReplicas is the number of Control Plane Machines created by the
+    // ControlPlaneSet controller which are ready.
+    ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+
+    // UpdatedReplicas is the number of non-terminated Control Plane Machines
+    // created by the ControlPlaneSet controller that have the desired
+    // provider spec.
+    UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
+}
```

Comment:
We'd probably want conditions to signal in progress scaling

Reply:
Yep fair point, I had intended to add them, one of my todos is to compare this API to the ControlPlane requirements from Cluster API to see if there's any guidelines there, with a view that this project may form a future implementation for IPI clusters

I had omitted them up to now because I don't know what types of conditions or reasons we might have to put on here, I'll add it in with a TODO for that.